Separate options for Wrapping func args and params

AnthonyMDev opened the pull request:
Resolves #421 
Allows you to set `--wrapparameters` option to override behavior of wrapping for function declarations to be different than other argument wrapping behavior.
This is fully tested.

nicklockwood commented on Sources/ParsingHelpers.swift:
```diff
@@ -876,9 +876,25 @@ extension Formatter {
                     }
+
+                    func isParameterList() -> Bool {
+                        if last(.keyword, before: i) == .keyword("func") {
```
This check is a bit fragile, e.g. it could return true in a case like:

```swift
func foo() { ... }
bar = (baz: 5, quux: 7)  // <- last keyword is func but these arguments aren't part of it
```

AnthonyMDev replied:
Hmm good point. I'm not sure how to make this better handle that case. Are there any other edge cases you can think of for this?

nicklockwood replied:
A check for a closing "}" between the keyword and the "(" would probably be sufficient.

AnthonyMDev replied:
Confirmed that fixes this. PR is coming soon. Thanks!

nicklockwood commented on Sources/ParsingHelpers.swift:
```diff
@@ -876,9 +876,25 @@ extension Formatter {
+                        }
+
+                        if last(.nonSpaceOrCommentOrLinebreak, before: i) == .delimiter(":"),
+                            next(.nonSpaceOrCommentOrLinebreak, after: endOfScope) == .operator("->", .infix) {
```
This doesn't handle `throws` or `rethrows` functions correctly, e.g.

```swift
let foo: (Int) throws -> String
```

You should be able to find some other cases in the rules that handle this more robustly. Probably worth extracting the logic into a ParsingHelper as it's not trivial.

AnthonyMDev replied:
Thanks for the comments! I'll re-review this and improve the solution tomorrow.

AnthonyMDev replied:
I can't think of anything else that could legally be between the closure parameter list and the `->` other than `throws` or `rethrows`. I can easily check for those. Can you think of anything else?

AnthonyMDev replied:
Could also have `@autoclosure` or `@escaping` before delimiter. Fixing that.

nicklockwood commented on Rules.md:
```diff
@@ -1514,13 +1514,40 @@ Align wrapped function arguments or collection elements.
-`--wraparguments` | Wrap function args: "before-first", "after-first", "preserve"
-`--wrapcollections` | Wrap array/dict: "before-first", "after-first", "preserve"
+`--wraparguments` | Wrap all arguments: "before-first", "after-first", "preserve" (default)
+`--wrapparameters` | Wrap function parameters: "before-first", "after-first", "preserve" (defaults to --wrapArguments value)
```
There's an 80-char limit for help text. It's fine to use abbreviations like `param` or `func` if necessary, and you can just say "preserve" is the default, since it's explained in more detail in the summary docs.

AnthonyMDev replied:
This is fixed. Will look into the other issues tomorrow.

AnthonyMDev commented:
@nicklockwood 
While building this, I've found that I think we should always `checkNestedScope` when wrapping arguments. In `ParsingHelpers.swift`, if the scope is `(`, we are turning `checkNestedScope` to `false`. I'm not sure why we are doing that. **When I set it to `true, all unit tests still pass**, so I'm not sure why we are setting it to `false` in the first place.

Consider the following example:
```swift
func foo(bar: Int, baz: (Int,
                         Bool, String) -> Int, quux: String) -> Int {}
```
With `checkNestedScope = false` this formats to:
```swift
func foo(bar: Int, baz: (
    Int,
    Bool, 
    String) -> Int, quux: String) -> Int {}
```
which I don't think makes much sense. If anything inside of the parameter list is wrapped, I think we should wrap the entire list, just like we do for `<` (generic arguments) and `[` (nested arrays & dictionaries). 

With `checkNestedScope = false` this formats much nicer to:
```swift
func foo(
    bar: Int,
    baz: (
        Int,
        Bool, 
        String
    ) -> Int,
    quux: String
) -> Int {}
```

My guess is that you added `checkNestedScope` when working on the `<` and `[` operator wrapping, and didn't consider that it should always be checking for nested scope when using `(` also.

Are you okay with making this change, or should I leave it with the current (which I think is awkward) behavior?

AnthonyMDev commented:
I've updated the PR to fix all of your concerns. I also removed the `checkNestedScope` so we are always checking nested scope now.

If that's not what you want, let me know and I'll change it back, but I strongly believe that checking nested scope on parameter/argument lists is the correct behavior.
